### DISCUSSION TITLE
Convert CRW to jpeg

### DISCUSSION THREAD

**Discussion by yuhaoy1994** [ImageMagick/ImageMagick#5880] [Help] [answered]
I am using this command to convert CRW file to jpeg image 
`magick convert -auto-orient my_crw.crw magick.jpg`, 
but found the result file 
![magick](https://user-images.githubusercontent.com/24372386/207626007-fbc87ec2-5cd5-4c1e-947e-74e0ffbe0d9d.jpg)

looks very different from the one I converted using the mac Preview 
![mac_preview](https://user-images.githubusercontent.com/24372386/207625909-350cd59e-eb35-4d7c-a76d-a5e3a06aecfd.jpg)

also I've tried to extract preview images directly from CRW with exiftool
`exiftool -JpgFromRaw -b  my_crw.crw > exiftool_preview.jpg`, and this gives result closer but still different to the Preview's result
![exiftool_preview](https://user-images.githubusercontent.com/24372386/207626665-bda2024e-f057-4073-a2b5-a82655257373.jpg). 

My question is am I missing anything with magick, why magick result looks so different?  

* my magick version info and I am running it inside a docker container of Debian GNU/Linux 11 (bullseye): 
> Version: ImageMagick 7.1.0-55 beta Q16-HDRI x86_64 f5cf5baad:20221210 https://imagemagick.org
Copyright: (C) 1999 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC HDRI OpenMP(4.5) 
Delegates (built-in): bzlib djvu fontconfig freetype heic jbig jng jp2 jpeg lcms lqr lzma openexr png raw tiff webp x xml zlib
Compiler: gcc (10.2)

CRW file
[RAW_CANON_10D.CRW.zip](https://github.com/ImageMagick/ImageMagick/files/10228859/RAW_CANON_10D.CRW.zip)

**Comment by snibgo** [comment, marked as answer]
The CRW file contains raw data from the camera's sensor. It also contains a JPEG that the camera has made from the raw data. The JPEG will have alterations according to the camera settings including white balance, and "prettifying" operations such as adjusting overall lightness and sharpening.

The JPEG is how the computer in the camera thinks the image should look. The raw data is available for you to make the image look as you want.

Your IM command processes the raw data, demosaicing it, and not doing much more. We often want to adjust white balance (WB) to be as the camera has calculated or set:
```
magick -define dng:use-camera-wb=true raw_canon_10d.crw out.tiff
```
See other "`dng:`" defines on https://imagemagick.org/script/defines.php

**Reply by yuhaoy1994** [reply]
Thanks for your explanation, appreciated. I am also wondering if it is possible to reproduce the mac Preview's result through magick. 

This IM command `magick -define dng:use-camera-wb=true raw_canon_10d.crw -auto-orient out.jpg` gives nice looking image result, but still much brighter. I've tried to use `-define dng:no-auto-bright=true` to disable histogram-based brightness change, and got this (the left side one below). It is a bit darker than the one I got with Preview (the right side one below)
![magick_2](https://user-images.githubusercontent.com/24372386/207747979-25ba158f-9b21-4e0c-8cc8-6784d38b76ee.jpg)![mac_preview](https://user-images.githubusercontent.com/24372386/207747811-7c960842-4170-4702-ab29-1fb335a15169.jpg)

It looks like how softwares perform brightness adjustment result in the difference between above two images, as you have mentioned 
> and "prettifying" operations such as adjusting overall lightness and sharpening

And if that is true, is there any gold standard that people usually perform to brightness change?

**Comment by fmw42** [comment]
Just a side note.  For IM 7 use magick, not magick convert.  All the other tools (identify, mogrify) need the magick preface, just not convert.  Convert was upgraded and the name changed to magick.  Magick convert works but is not the same as magick.

**Comment by snibgo** [comment]
>  I am also wondering if it is possible to reproduce the mac Preview's result through magick.

I don't know if the mac Preview processes the raw sensor image, or extracts the embedded JPEG, and I don't know what processing the Preview does.

When you have two versions of the same image with different colours or lightnesses, you can adjust one image to (roughly) match the other. For example, see [Setting the mean](http://im.snibgo.com/setmean.htm).

> And if that is true, is there any gold standard that people usually perform to brightness change?

Not really. A typical photograph in a display colorspace such as sRGB or Adobe1998 will have an overall average lightness of around 0.5, on a scale of 0.0 to 1.0.

Extracting the jpeg from your raw file, then:
```
magick e.jpg -verbose info:
```
... includes this:
```
  Image statistics:
    Overall:
      min: 0  (0)
      max: 255 (1)
      mean: 71.8219 (0.281655)
```
The overall mean, on a scale of 0.0 to 1.0, is 0.281655. And the image does look dark to me.

Lightness is a matter of personal preference, and subject matter, and mood and emotions and other factors.

**Reply by yuhaoy1994** [reply]
Well, at this point, I tend to believe that Preview has some"default processes" with hand-craft parameters to be applied to images before it is exported as JPEG and the processes might vary to different softwares. So we will probably get slight different views of the same raw file, when it is opened with Preview, Lightroom and etc.   
It feels like raw file processing is very subjective, and I shouldn't bother to reproducing the Preview's result.

**Comment by snibgo** [comment]
> It feels like raw file processing is very subjective, ...

Yes. It can be fairly objective, if we simply convert the demosaiced image to the colorspace of your output device, and do no other processing. But we often prefer to do some processing, so the image more closely says what we want it to say.
